### PR TITLE
Split-up lib.rs

### PR DESCRIPTION
**Pull request by regexident**
This PR splits up the `lib.rs` file, introducing separate files for:

- `Relation<T>` -> `relation.rs`
- `Iteration` -> `iteration.rs`
- `Variable<T>` -> `variable.rs`

Splitting up the single 555loc file required me to make a select few previously `private` bindings `pub(crate)`, instead.
However I think overall this greatly improves the accessibility of the project and is worth it.

---

This is part #1 of a stacked pull-request:

```
└── Split-up\ lib.rs 👈🏻
    └── Cleanup\ generics (#33)
        └── Support arbitrary tuples (#34)
```

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
Hi @regexident!

I'm in full agreement with this PR, however I think the only active person with r+ rights in this repo is @nikomatsakis, and they are quite busy. I'll try to convince them to give either @lqd or me (preferably both) permission to approve PRs so we can get things moving.

File: src/relation.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
+use std::iter::FromIterator;
+
+use crate::{
+    join::{self},
```
Comment (ecstatic-morse) [inline src/relation.rs:4]:
```suggestion
    join,
```

**Review by ecstatic-morse — COMMENTED** [review]
Just one small nit and then this is good to go. Thank you!

**Comment by ecstatic-morse** [comment]
I have some stuff in the pipeline that will conflict with this, so I went ahead and added the nit on top before merging (I would have rebased but it's non-trivial to preserve authorship info this way). Thanks again @regexident.

**Comment by regexident** [comment]
Thanks @ecstatic-morse!

**Pull request merged**